vllm/core/block_manager.py

```diff
@@ -1,12 +1,61 @@
 """A block manager that manages token blocks."""
 import enum
+from itertools import takewhile
+from os.path import commonprefix
+from time import monotonic
 from typing import Dict, List, Optional, Set, Tuple
 
 from vllm.block import BlockTable, PhysicalTokenBlock
 from vllm.sequence import Sequence, SequenceGroup, SequenceStatus
 from vllm.utils import Device
 
 
+class EvictionPolicy(enum.Enum):
+    """Enum for eviction policy used by BlockAllocator."""
+    LRU = enum.auto()
+
+
+def lru_eviction(
+        free_table: Dict[int, PhysicalTokenBlock]) -> PhysicalTokenBlock:
+    free_blocks: List[PhysicalTokenBlock] = list(free_table.values())
+    if len(free_blocks) == 0:
+        raise ValueError("No usable cache memory left")
+
+    # Find lowest timestamp
+    lowest_timestamp = monotonic()
+    for block in free_blocks:
+        if block.last_accessed < lowest_timestamp:
+            lowest_timestamp = block.last_accessed
+
+    # Find all blocks with the lowest timestamp
+    least_recent: List[PhysicalTokenBlock] = []
+    for block in free_blocks:
+        if block.last_accessed == lowest_timestamp:
+            least_recent.append(block)
+
+    # Find highest prefix count per block
+    highest_prefix_count = 0
+    for block in least_recent:
+        if block.prefix_len > highest_prefix_count:
+            highest_prefix_count = block.prefix_len
+
+    # Find all blocks with the lowest timestamp
+    eviction_candidates: List[PhysicalTokenBlock] = []
+    for block in least_recent:
+        if block.prefix_len == highest_prefix_count:
+            eviction_candidates.append(block)
+
+    # Arbitrarily evict the first candidate
+    if len(eviction_candidates) == 0:
+        raise ValueError("No usable cache memory left")
+
+    evicted_block = eviction_candidates[0]
+    del free_table[evicted_block.block_hash]
+
+    evicted_block.computed = False
+    return evicted_block
```

Comment:
This is just a nit: Since we're only returning the first candidate anyways, should we evict as soon as we have a match for `block.prefix_len == highest_prefix_count` instead of going through the whole list?